File: Proxy/c#/Proxy.cs
interface ISite
{
    string GetPage(int num);
}

class Site : ISite
{
    public string GetPage(int num) => "page " + num;
}

class SiteProxy : ISite
{
    private ISite _site;
    private Dictionary<int, string> _cache;

    public SiteProxy(ISite site)
    {
        _site = site;
        _cache = new Dictionary<int, string>();
    }

    public string GetPage(int num)
    {
        string page;
        if (_cache.ContainsKey(num))
        {
            page = _cache[num];
            page = "from cache " + page;
        }
        else
        {
            page = _site.GetPage(num);
            _cache.Add(num, page);
        }

        return page;
    }
}

class Program
{
    static void Main()
    {
        ISite mySite = new SiteProxy(new Site());
        
        Console.WriteLine(mySite.GetPage(1));
        Console.WriteLine(mySite.GetPage(2));
        Console.WriteLine(mySite.GetPage(3));
        
        Console.WriteLine(mySite.GetPage(2));
    }
}

File: Proxy/java/Proxy.java
import java.util.HashMap;
import java.util.Map;

interface ISite{
    String getPage(int num);
}

class Site implements ISite{
    public String getPage(int num){
        return "page " + num;
    }
}

class SiteProxy implements ISite{
    private ISite site;
    
    private Map<Integer, String> cache;

    public SiteProxy(ISite site){
        this.site = site;
        cache = new HashMap<>();
    }

    public String getPage(int num){
        String page;
        if (cache.containsKey(num)){
            page = cache.get(num);
            page = "from cache " + page;
        }
        else{
            page = site.getPage(num);
            cache.put(num, page);
        }
        return page;
    }
}

public class Main{
    public static void main(String[] args){
        ISite mySite = new SiteProxy(new Site());

        System.out.println(mySite.getPage(1));
        System.out.println(mySite.getPage(2));
        System.out.println(mySite.getPage(3));

        System.out.println(mySite.getPage(2));
    }
}